File: static/js/benchmark/detourLimits.js
/**
 * Detour policy for smart routes (browser benchmark + aligned with API).
 * - Max extra distance vs direct baseline: 5 km
 * - If a candidate beats the best within 5 km on env score, allow up to +500 m more detour (5.5 km total)
 */

export const MAX_DETOUR_M = 5000;
export const BETTER_SCORE_ENV_TOLERANCE_M = 500;

export function detourMetres(routeDistanceM, baselineDistanceM) {
    const base = Number(baselineDistanceM) || 0;
    const route = Number(routeDistanceM) || 0;
    return Math.max(0, route - base);
}

/**
 * Pick best candidate by environmental score (higher = better).
 * Each candidate: { distance_m, env_score, ... }
 */
const ENV_SCORE_EPSILON = 0.02;
const MIN_MEANINGFUL_DETOUR_M = 80;

/**
 * True if candidate could improve EI vs direct (not just a tie on the same path).
 */
export function isMeaningfulImprovement(candidate, baselineEnvScore) {
    if (baselineEnvScore == null || candidate?.env_score == null) return true;
    if (candidate.env_score > baselineEnvScore + ENV_SCORE_EPSILON) return true;
    if (
        candidate.detour_m >= MIN_MEANINGFUL_DETOUR_M &&
        candidate.env_score >= baselineEnvScore - ENV_SCORE_EPSILON
    ) {
        return true;
    }
    return false;
}

export function selectBestWithinDetourPolicy(
    candidates,
    baselineDistanceM,
    baselineEnvScore = null,
) {
    if (!candidates?.length || !baselineDistanceM) {
        return { chosen: null, reason: 'no_candidates' };
    }

    const withDetour = candidates.map((c) => ({
        ...c,
        detour_m: detourMetres(c.distance_m, baselineDistanceM),
    }));

    let pool = withDetour;
    if (baselineEnvScore != null) {
        const improving = withDetour.filter((c) =>
            isMeaningfulImprovement(c, baselineEnvScore),
        );
        if (improving.length > 0) {
            pool = improving;
        } else {
            return { chosen: null, reason: 'no_env_improvement_over_direct' };
        }
    }

    const withinHard = pool.filter((c) => c.detour_m <= MAX_DETOUR_M);
    const bestHard =
        withinHard.length > 0
            ? withinHard.reduce((a, b) => ((a.env_score ?? -1) >= (b.env_score ?? -1) ? a : b))
            : null;

    const hardScore = bestHard?.env_score ?? -Infinity;
    const softCap = MAX_DETOUR_M + BETTER_SCORE_ENV_TOLERANCE_M;
    const withinSoft = pool.filter(
        (c) =>
            c.detour_m <= softCap &&
            c.detour_m > MAX_DETOUR_M &&
            typeof c.env_score === 'number' &&
            c.env_score > hardScore,
    );

    if (withinSoft.length > 0) {
        const chosen = withinSoft.reduce((a, b) => (a.env_score >= b.env_score ? a : b));
        return {
            chosen,
            reason: 'soft_cap_better_env',
            max_detour_m: softCap,
            best_hard_detour_m: bestHard?.detour_m ?? null,
        };
    }

    if (bestHard) {
        return {
            chosen: bestHard,
            reason: 'within_5km_detour',
            max_detour_m: MAX_DETOUR_M,
        };
    }

    const withinSoftCap = pool.filter((c) => c.detour_m <= softCap);
    if (withinSoftCap.length > 0) {
        const chosen = withinSoftCap.reduce((a, b) =>
            (a.env_score ?? -1) >= (b.env_score ?? -1) ? a : b,
        );
        return {
            chosen,
            reason: 'within_5_5km_best_env',
            max_detour_m: softCap,
        };
    }

    return {
        chosen: null,
        reason: 'all_exceed_detour_cap',
        max_detour_m: softCap,
    };
}

File: evaluations/detour_limits.py
"""Detour limits for smart-route selection (API benchmark, mirrors browser policy)."""

from typing import Any, Dict, List, Optional

MAX_DETOUR_M = 5000
BETTER_SCORE_ENV_TOLERANCE_M = 500


def detour_metres(route_distance_m: float, baseline_distance_m: float) -> float:
    return max(0.0, float(route_distance_m or 0) - float(baseline_distance_m or 0))


def select_best_within_detour_policy(
    candidates: List[Dict[str, Any]],
    baseline_distance_m: float,
) -> Dict[str, Any]:
    if not candidates or not baseline_distance_m:
        return {'chosen': None, 'reason': 'no_candidates'}

    enriched = []
    for c in candidates:
        d_m = float(c.get('distance_m') or 0)
        enriched.append({**c, 'detour_m': detour_metres(d_m, baseline_distance_m)})

    within_hard = [c for c in enriched if c['detour_m'] <= MAX_DETOUR_M]
    best_hard = (
        max(within_hard, key=lambda c: float(c.get('env_score') or -1))
        if within_hard
        else None
    )
    hard_score = float(best_hard['env_score']) if best_hard else float('-inf')
    soft_cap = MAX_DETOUR_M + BETTER_SCORE_ENV_TOLERANCE_M
    within_soft = [
        c
        for c in enriched
        if c['detour_m'] <= soft_cap
        and c['detour_m'] > MAX_DETOUR_M
        and float(c.get('env_score') or -1) > hard_score
    ]

    if within_soft:
        chosen = max(within_soft, key=lambda c: float(c.get('env_score') or -1))
        return {
            'chosen': chosen,
            'reason': 'soft_cap_better_env',
            'max_detour_m': soft_cap,
        }

    if best_hard:
        return {
            'chosen': best_hard,
            'reason': 'within_5km_detour',
            'max_detour_m': MAX_DETOUR_M,
        }

    within_soft_cap = [c for c in enriched if c['detour_m'] <= soft_cap]
    if within_soft_cap:
        chosen = max(within_soft_cap, key=lambda c: float(c.get('env_score') or -1))
        return {
            'chosen': chosen,
            'reason': 'within_5_5km_best_env',
            'max_detour_m': soft_cap,
        }

    return {
        'chosen': None,
        'reason': 'all_exceed_detour_cap',
        'max_detour_m': soft_cap,
    }
